src/java.base/share/classes/java/util/stream/Gatherers.java

```diff
@@ -0,0 +1,596 @@
+     * }
+     *
+     * @param windowSize the size of the windows
+     * @param <TR> the type of elements the returned gatherer consumes
+     *             and the contents of the windows it produces
+     * @return a new gatherer which groups elements into fixed-size windows
+     * @throws IllegalArgumentException when groupSize is less than 1
+     */
+    public static <TR> Gatherer<TR, ?, List<TR>> windowFixed(int windowSize) {
+        if (windowSize < 1)
+            throw new IllegalArgumentException("'groupSize' must be greater than zero");
+
+        return Gatherer.ofSequential(
+                // Initializer
+                () -> new ArrayList<>(windowSize),
+
+                // Integrator
+                Integrator.<ArrayList<TR>,TR,List<TR>>ofGreedy((window, e, downstream) -> {
+                    window.add(e);
+                    if (window.size() < windowSize) {
+                        return true;
+                    } else {
+                        var full = List.copyOf(window);
+                        window.clear();
+                        return downstream.push(full);
+                    }
+                }),
+
+                // Finisher
+                (window, downstream) -> {
+                    if(!downstream.isRejecting() && !window.isEmpty())
+                        downstream.push(List.copyOf(window));
+                }
+        );
+    }
+
+    /**
+     * Gathers elements into sliding windows, sliding out the most previous
+     * element and sliding in the next element for each subsequent window.
+     * If the stream is empty then no window will be produced. If the size of
+     * the stream is smaller than the window size then only one window will
+     * be emitted, containing all elements.
+     *
+     * <p>Example:
+     * {@snippet lang = java:
+     * // will contain: [[1, 2], [2, 3], [3, 4], [4, 5], [5, 6], [6, 7], [7, 8]]
+     * List<List<Integer>> windows =
+     *     Stream.of(1,2,3,4,5,6,7,8).gather(Gatherers.windowSliding(2)).toList();
+     * }
+     *
+     * @param windowSize the size of the windows
+     * @param <TR> the type of elements the returned gatherer consumes
+     *             and the contents of the windows it produces
+     * @return a new gatherer which groups elements into sliding windows
+     * @throws IllegalArgumentException when windowSize is less than 1
+     */
+    public static <TR> Gatherer<TR, ?, List<TR>> windowSliding(int windowSize) {
+        if (windowSize < 1)
+            throw new IllegalArgumentException("'groupSize' must be greater than zero");
+
+        @SuppressWarnings("serial")
+        class SlidingWindow extends ArrayDeque<TR> {
+            boolean firstWindow = true;
+        }
+        return Gatherer.ofSequential(
+                // Initializer
+                SlidingWindow::new,
+
+                // Integrator
+                Integrator.ofGreedy((window, e, downstream) -> {
+                    window.addLast(e);
```

Comment:
Too bad it will reject null elements in the current implementation. This should be documented, or better fixed. We can use ArrayList and keep track of the current start index, and export slides manually using arraycopy twice and  `listFromTrustedArrayNullsAllowed`. It's only a little bit trickier and should not be slower than the current implementation. I can write it if you don't want to do it by yourself :-)

Reply:
Took a stab at reimplementing both windowFixed and windowSliding to allow null elements in the stream: https://github.com/openjdk/jdk/pull/16420/commits/57fea44ce71ae94107f3fb6d7f104d35839371d9